File: javascript/robot-name/robot-name.js
const ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const usedNames = new Set();

export class Robot {
    #name;

    constructor() {
        this.#name = this.#generateName();
    }

    get name() {
        return this.#name;
    }

    reset() {
        this.#name = this.#generateName();
    }

    #generateName() {
        let newName = '';

        do {

            let letters = ALPHABET[Math.floor(Math.random() * 26)] 
                + ALPHABET[Math.floor(Math.random() * 26)];

            newName = `${letters}${100 + Math.floor(Math.random() * 900)}`;
        } while (usedNames.has(newName));

        usedNames.add(newName);
        return newName;
    }

}


Robot.releaseNames = () => usedNames.clear();

File: clojure/robot-name/src/robot_name.clj
(ns robot-name)

(defn generate-name []
  (def alphabet "ABCDEFGHIJKLMNOPQRSTUVWXYZ") 
  (str 
    (get alphabet (rand-int 27)) 
    (get alphabet (rand-int 27))
    (+ 101 (rand-int 899))
  )

)

(defn robot [] 
  (def new-name (generate-name))
  (atom {:name new-name})
)


(defn robot-name [robot]
  (:name @robot)
)

(defn reset-name [robot]
  (swap! robot assoc :name (generate-name))
)
